frontend/src/components/app-config/ai-config.tsx

```diff
@@ -1063,22 +1117,204 @@ export const AiModelDisplayConfig: React.FC<AiConfigProps> = ({
           className="flex-1 overflow-auto outline-none focus-visible:outline-none"
           selectionMode="none"
         >
-          {[...modelsByProvider.entries()].map(([providerId, models]) => (
+          {listModelsByProvider.map(([providerId, models]) => (
             <ProviderTreeItem
               key={providerId}
               providerId={providerId}
               models={models}
               enabledModels={currentDisplayedModelsSet}
               onToggleModel={toggleModelDisplay}
               onToggleProvider={toggleProviderModels}
+              onDeleteModel={deleteModel}
             />
           ))}
         </Tree>
       </div>
+      <AddModelForm form={form} customModels={customModels} />
     </SettingGroup>
   );
 };
 
+export const AddModelForm: React.FC<{
+  form: UseFormReturn<UserConfig>;
+  customModels: QualifiedModelId[];
+}> = ({ form, customModels }) => {
+  const [isFormOpen, setIsFormOpen] = useState(false);
+  const [modelAdded, setModelAdded] = useState(false);
+  const [provider, setProvider] = useState<ProviderId | "custom" | null>(null);
+  const [customProviderName, setCustomProviderName] = useState("");
+  const [modelName, setModelName] = useState("");
+
+  const providerSelectId = useId();
+  const customProviderInputId = useId();
+  const modelNameInputId = useId();
+
+  const isCustomProvider = provider === "custom";
+  const providerName = isCustomProvider ? customProviderName : provider;
+  const hasValidValues = providerName?.trim() && modelName?.trim();
+
+  const resetForm = () => {
+    setProvider(null);
+    setCustomProviderName("");
+    setModelName("");
+    setIsFormOpen(false);
+  };
+
+  const handleAddModel = () => {
+    if (!hasValidValues) {
+      return;
+    }
+
+    const newModel = new AiModelId(
+      providerName as ProviderId,
+      modelName as ShortModelId,
+    );
+
+    form.setValue("ai.models.custom_models", [newModel.id, ...customModels]);
+    resetForm();
+
+    // Show model added message for 2 seconds
+    setModelAdded(true);
+    setTimeout(() => setModelAdded(false), 2000);
+  };
+
+  const providerClassName = "w-40 truncate";
+
+  const providerSelect = (
+    <div className="flex flex-col gap-2">
+      <div className="flex items-center gap-2">
+        <Label
+          htmlFor={providerSelectId}
+          className="text-sm font-medium text-muted-foreground min-w-12"
+        >
+          Provider
+        </Label>
+        <Select
```

Comment:
the dropdown is nice, but i assume most of the custom models will be Custom anyways. thoughts on just the `provider` text input? only if you think this code is fragile / extra maintenance cost